include/boost/gil/image_processing/scaling.hpp

```diff
@@ -0,0 +1,115 @@
+#include <boost/gil/image_view.hpp>
+#include <boost/gil/rgb.hpp>
+#include <boost/gil/pixel.hpp>
+#include <boost/gil/image_processing/numeric.hpp>
+
+namespace boost{ namespace gil{
+/// \defgroup ScalingAlgorithms
+/// \brief Algorthims suitable for rescaling
+///
+/// These algorithms are used to improve image
+/// quality after image resizing is made.
+
+/// \defgroup DownScalingAlgorithms
+/// \ingroup ScalingAlgorithms
+/// \brief Algorthims suitable for downscaling
+///
+/// These algorithms provide best results when used
+/// for downscaling. Using for upscaling will probably
+/// provide less than good results.
+
+
+/// \brief a single step of lanczos downscaling
+/// \ingroup DownScalingAlgorithms
+///
+/// Use this algorithm to scale down source image
+/// into a smaller image with reasonable quality.
+/// Do note that having a look at the output once
+/// is a good idea, since it might have ringing
+/// artifacts.
+template <typename ImageView>
+void lanczos_at(
+    long int source_x, 
+    long int source_y, 
+    long int target_x, 
+    long int target_y, 
+    ImageView input_view, 
+    ImageView output_view, 
+    long int a) 
+{
+    using pixel_t = typename std::remove_reference<
+                      decltype(std::declval<ImageView>()(0, 0))
+                    >::type;
+    // C++11 doesn't allow auto in lambdas
+    using channel_t = typename std::remove_reference<
+                        decltype(
+                            std::declval<pixel_t>().at(
+                                std::integral_constant<int, 0>{}
+                            )
+                        )
+                       >::type;
+    pixel_t result_pixel;
+    boost::gil::static_transform(result_pixel, result_pixel, 
+                                 [](channel_t) {
+                                     return static_cast<channel_t>(0);
+                                    });
+    
+    for (long int y_i = std::max(source_y - a + 1, 0l); 
+         y_i <= std::min(source_y + a, input_view.height() - 1); 
```

Comment:
@simmplecoder I think you need some casts here.

@stefanseefeld Here comes my rant 😄  If `ptrdiff_t` was used consistently across the GIL code base and this code, this line would have not failed to compile due to `scaling.hpp(58): error C2672: 'std::min': no matching overloaded function found`

Reply:
So `- 1` is promoting unsigned to signed? Not sure if it is called promotion though

Reply:
Is there a piece of long int arithmetics that I don't understand? I believe I'm not mixing types, except that + 1 and -1.

Reply:
This is more about operand types mismatch while there is single `T` expected. For visual example:

```
#include <algorithm>
#include <cstddef>
#include <type_traits>

template <typename T>
void foo(T const& a, T const& b) {}

int main()
{
    long int a;
    std::ptrdiff_t b;

    static_assert(!std::is_same<long int, std::ptrdiff_t>::value);
    static_assert(std::is_same<decltype(b + 1), std::ptrdiff_t>::value);
    static_assert(std::is_same<decltype(a + a), long int>::value);

    std::min(a, b);    // T=long != T=int or T=long != T=long long
    foo(a + a, b - 1); // equivalent of std::min/max case
}
```

Reply:
Could you please tell to what exactly? I'm still not getting what is going wrong. I read integer promotion page on cppref completely but it doesn't say anything about this. I looked at the error messages, but I am not sure if it is replacing `std::ptrdiff_t` with direct type.

Reply:
- the operands are of different types, while `std::min` requires both are of the same type `T`
- you need to cast either
- prefer to cast to the one larger capacity i.e.  `b` is a type of size `>=` of type of `a`, so you need

```
std::min(static_cast<std::ptrdiff_t>(a), b);
foo(static_cast<std::ptrdiff_t>(a + a), b - 1);
```

In other words, since `height()` and `width()` return `ptrdiff_t`, I insist to ensure that the other operands are unified to `ptrdiff_t`. 

I also pointed that out during discussion with @stefanseefeld, consistent use of types does matter and helps to avoid sitautions like above. Even if the `ptrdiff_t` in particular is not entirely correct choice, it is a good choice given the current contenxt of the current GIL interface (i.e. `height()` and `width()` , etc.)

Reply:
Sorry for bothering you even more, but could you please explain where did `std::ptrdiff_t` come from? Last time I checked the return type of `width()` and `height()`  was `long int`, am I wrong?

Reply:
I see, I missed the PR that uses `std::ptrdiff_t`, sorry.

Reply:
Wherever you usee `x_coord_t` or `y_coord_t`, those are most likely `ptrdiff_t` as they are derived from iterator's `difference_type` which in both, standard library and Boost.Iterator, default to `ptrdiff_t`. So, these are `ptrdiff_t`:

```
    x_coord_t        width()                 const { return dimensions().x; }
    y_coord_t        height()                const { return dimensions().y; }
```

Then, GIL sets default `point_t` as `point<ptrdiff_t>`, so  this is `ptrdiff_t` as well:

```
    const point_t&   dimensions()            const { return _dimensions; }
```

> Sorry for bothering you even more

No need. You ask valid questions.

Reply:
I see, thanks! I believe I addressed all issues, including the lambda formatting

Reply:
Hm, it seems like that was not enough. Should I try to cast both sides? Here is new version:

    std::min(static_cast<std::ptrdiff_t>(source_x + a), input_view.width() - 1)

I believe I casted the right argument

Reply:
@simmplecoder I've taken the liberty to `ptrdiff_t`-improve your code. Here is my `git diff` output for your branch with my modifications:

```
diff --git a/include/boost/gil/image_processing/scaling.hpp b/include/boost/gil/image_processing/scaling.hpp
index 562486f27..24b91b2a0 100644
--- a/include/boost/gil/image_processing/scaling.hpp
+++ b/include/boost/gil/image_processing/scaling.hpp
@@ -29,13 +29,13 @@ namespace boost{ namespace gil{
 /// artifacts.
 template <typename ImageView>
 void lanczos_at(
-    long int source_x, 
-    long int source_y, 
-    long int target_x, 
-    long int target_y, 
+    std::ptrdiff_t source_x, 
+    std::ptrdiff_t source_y, 
+    std::ptrdiff_t target_x, 
+    std::ptrdiff_t target_y, 
     ImageView input_view, 
     ImageView output_view, 
-    long int a) 
+    std::ptrdiff_t a) 
 {
     using pixel_t = typename std::remove_reference<
                       decltype(std::declval<ImageView>()(0, 0))
@@ -51,13 +51,13 @@ void lanczos_at(
     pixel_t result_pixel;
     boost::gil::static_transform(result_pixel, result_pixel, 
         [](channel_t) { return static_cast<channel_t>(0); });
-    
-    for (long int y_i = std::max(source_y - a + 1l, 0l); 
-         y_i <= std::min(static_cast<std::ptrdiff_t>(source_y + a), input_view.height() - 1l);
+
+    for (std::ptrdiff_t y_i = std::max(source_y - a + 1, std::ptrdiff_t{0}); 
+         y_i <= std::min(source_y + a, input_view.height() - 1);
          ++y_i) 
         {
-        for (long int x_i = std::max(source_x - a + 1l, 0l); 
-             x_i <= std::min(static_cast<std::ptrdiff_t>(source_x + a), input_view.width() - 1l); 
+        for (std::ptrdiff_t x_i = std::max(source_x - a + 1, std::ptrdiff_t{0}); 
+             x_i <= std::min(source_x + a, input_view.width() - 1); 
              ++x_i) 
         {
             double lanczos_response = boost::gil::lanczos(source_x - x_i, a) 
@@ -88,16 +88,16 @@ void lanczos_at(
 /// https://en.wikipedia.org/wiki/Lanczos_resampling
 /// with standardinzed cardinal sin (sinc)
 template <typename ImageView>
-void scale_lanczos(ImageView input_view, ImageView output_view, long int a) 
+void scale_lanczos(ImageView input_view, ImageView output_view, std::ptrdiff_t a) 
 {
     double scale_x = (static_cast<double>(output_view.width())) 
                      / static_cast<double>(input_view.width());
     double scale_y = (static_cast<double>(output_view.height())) 
                      / static_cast<double>(input_view.height());
 
-    for (long int y = 0; y < output_view.height(); ++y) 
+    for (std::ptrdiff_t y = 0; y < output_view.height(); ++y) 
     {
-        for (long int x = 0; x < output_view.width(); ++x) 
+        for (std::ptrdiff_t x = 0; x < output_view.width(); ++x) 
         {
             boost::gil::lanczos_at(
                 x / scale_x, 
```

The integer suffixes also introduced some inconsistencies depending on target architecture.

It compiles for me using VS2019 for 32/64 bit.

Reply:
One more comment, the patch above also illustrates the signature could read

- `void lanczos_at(x_coord_t src_x, y_coord_t src_y, x_coord_t dst_x, y_coord_t dst_y, ...)`
- `void lanczos_at(point_t src, point_t dst,...)`

That is a not uncommon pattern used in GIL code. That is also why maintaining the consistency of the type representing image or view dimensions, row and pixel indexes, etc. does pay off.